### PR TITLE
added a micro kernel

### PR DESCRIPTION
**Pull request by fabpot**
| Q | A |
| --- | --- |
| Bug fix? | no |
| New feature? | yes |
| BC breaks? | no |
| Deprecations? | no |
| Tests pass? | yes |
| Fixed tickets | n/a |
| License | MIT |
| Doc PR | n/a |

Related to #15948 and #15820

### CODE REVIEW

File: src/Symfony/Component/HttpKernel/MicroKernel.php [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+use Symfony\Component\Config\Resource\FileResource;
+use Symfony\Component\DependencyInjection\ContainerBuilder;
+
+/**
+ * A Kernel that provides configuration hooks.
+ *
+ * @author Ryan Weaver <ryan@knpuniversity.com>
+ * @author Fabien Potencier <fabien@symfony.com>
+ */
+abstract class MicroKernel extends Kernel
```
Comment (fabpot) [inline src/Symfony/Component/HttpKernel/MicroKernel.php]: `MicroKernel` is probably not the right class name, it's just a regular Kernel with some method helpers.

**Comment by weaverryan** [comment]
Obviously, there are a few minor todos as you mentioned, but this accomplishes all of the goals of my other PR's that you listed, but indeed, in a much simpler way (and a smaller code footprint). I'm very pleased by this version!

File: src/Symfony/Component/HttpKernel/MicroKernel.php [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+                        return $this->configureRoutes(new RouteCollectionBuilder($loader))->build();
+                    },
+                    'type' => 'closure',
+                ),
+            ));
+
+            $this->configureExtensions($container, $loader);
+            $this->configureServices($container, $loader);
+
+            $loader->addResource(new FileResource(__FILE__));
```
Comment (weaverryan) [inline src/Symfony/Component/HttpKernel/MicroKernel.php]: I think we need this to include the sub-class:

```php
$container->addObjectResource($this);
```

File: src/Symfony/Component/HttpKernel/MicroKernel.php (hunk 1 above)
Comment (weaverryan) [inline src/Symfony/Component/HttpKernel/MicroKernel.php]: Hmm, yes. Will we need to split this class into 2 parts (with service stuff in HttpKernel and route stuff in FrameworkBundle)?

Comment (weaverryan) [inline src/Symfony/Component/HttpKernel/MicroKernel.php]: I can't think of a good name, so let me throw out some not great names for brainstorming:

- ConfigurableKernel
- CompleteKernel
- AutonomousKernel
- BuildableKernel

File: src/Symfony/Component/HttpKernel/MicroKernel.php [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+
+        parent::boot();
+
+        // 3.0 only
+        $this->getContainer()->get('router')->setResource(function () {
+            $loader = $this->getContainerLoader($this->getContainer());
+
+            return $this->configureRoutes(new RouteCollectionBuilder($loader))->build();
+        });
+    }
```
Comment (weaverryan) [inline src/Symfony/Component/HttpKernel/MicroKernel.php]: This starts to feel a bit worse than the ServiceRouteLoader approach, which also had some supporters for other use-cases. If there were a way to pull this off in a nice way without any new code, that would be great. But otherwise, I like #15742

Comment (weaverryan) [inline src/Symfony/Component/HttpKernel/MicroKernel.php]: I updated that PR, and have a suggestion that would remove the need for putting the `RouteLoaderInterface` on the kernel: https://github.com/symfony/symfony/pull/15742/files#r40755658

File: src/Symfony/Component/HttpKernel/MicroKernel.php (hunk 1 above)
Comment (javiereguiluz) [inline src/Symfony/Component/HttpKernel/MicroKernel.php]: I love `MicroKernel` name. I cannot think of a better alternative :)

Comment (fabpot) [inline src/Symfony/Component/HttpKernel/MicroKernel.php]: It's not about a sexy name, but having a descriptive one; there is nothing Micro here.

Comment (weaverryan) [inline src/Symfony/Component/HttpKernel/MicroKernel.php]: I like:

* FrameworkKernel (since, we define a framework as at least allowing for services+routes+controllers)
* ConfigurableKernel
* SelfContainedKernel

**Comment by weaverryan** [comment]
The updated routing method looks good to me. :+1: in general - except for the name of course :)

File: src/Symfony/Component/HttpKernel/MicroKernel.php [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+            ));
+
+            $this->configureExtensions($container, $loader);
+            $this->configureServices($container, $loader);
+
+            $container->addObjectResource($this);
+        });
+    }
+
+    private function loadRoutes(LoaderInterface $loader)
```
Comment (weaverryan) [inline src/Symfony/Component/HttpKernel/MicroKernel.php]: I think this will need to be public

**Comment by weaverryan** [comment]
@fabpot if we want some tests (there's not much to test, but there could be some) or other changes and I can help - let me know and I can steal this PR and make those changes (and take all the credit!)

File: src/Symfony/Component/HttpKernel/MicroKernel.php (hunk 1 above)
Comment (ogizanagi) [inline src/Symfony/Component/HttpKernel/MicroKernel.php]: - StandaloneKernel
- SelfSustainedKernel
- SelfStandingKernel
- AdvancedKernel

Comment (weaverryan) [inline src/Symfony/Component/HttpKernel/MicroKernel.php]: StandaloneKernel is decent :)

Comment (javiereguiluz) [inline src/Symfony/Component/HttpKernel/MicroKernel.php]: The problem with "stand-alone" is that it's spelled with a dash, which will make people think: should I spell it as `StandaloneKernel` or `StandAloneKernel`. Besides, the dictionary describes "stand-alone" as: *"able to operate without control from another system"*. Is that entirely true in this case?

**Comment by fabpot** [comment]
Here is a small usage example:

```php
<?php

require_once __DIR__.'/../vendor/autoload.php';

use Symfony\Component\Config\Loader\LoaderInterface;
use Symfony\Component\Debug\Debug;
use Symfony\Component\DependencyInjection\ContainerBuilder;
use Symfony\Component\HttpFoundation\Request;
use Symfony\Component\HttpFoundation\Response;
use Symfony\Component\HttpKernel\MicroKernel;
use Symfony\Component\Routing\RouteCollectionBuilder;

class AppKernel extends MicroKernel
{
    public function home()
    {
        return new Response('Foo');
    }

    public function registerBundles()
    {
        return array(
            new Symfony\Bundle\FrameworkBundle\FrameworkBundle(),
        );
    }

    protected function configureRoutes(RouteCollectionBuilder $routes)
    {
        $routes->add('/', 'kernel:home');
    }

    protected function configureExtensions(ContainerBuilder $c, LoaderInterface $loader)
    {
    }

    protected function configureServices(ContainerBuilder $c, LoaderInterface $loader)
    {
    }
}

$kernel = new AppKernel('dev', true);
$request = Request::createFromGlobals();
$response = $kernel->handle($request);
$response->send();
$kernel->terminate($request, $response);
```

I'm not quite satisfied as we cannot use a closure for controllers (as Symfony dumps the routes as a PHP file).

**Comment by fabpot** [comment]
Question: is it really something we want to package in Symfony itself. As you can see, the code is quite minimal, so isn't it something for a `small/micro` distribution? Or part of a tutorial in the docs? For me, that's more like a starting point rather than some kind of infrastructure that needs to be standardized.

**Comment by xabbuh** [comment]
If this is something that will be asked for frequently, I would say yes (personally I don't see any use cases for myself right now). The code that would reside inside the core looks good so far.

**Comment by fabpot** [comment]
The fact that closures cannot be used defeats the initial purpose (at least for me). With closure support, that would make it something interesting for small websites, APIs, things that Silex is good at.

**Comment by weaverryan** [comment]
@fabpot I want the code *somewhere* where people can just use it, so people can get a minimal Symfony project up and running really quickly and with minimal setup. If we make the user do that boilerplate, it defeats the purpose.

My vote is for core because I want say: "you can create a symfony app by requiring `symfony/symfony` and creating a single file that extends MicroKernel" (or whatever we call it).

Also, I think the kernel should be used in the SE: I think loads things in a way that's more clear, but basically works the same (you could easily remove `routing_dev.yml` for example).

And finally, yes, Silex is awesome. But it doesn't have bundle support, which means I forfeit a lot of solutions I might be accustomed to using. A Silex app also can't evolve into a full Symfony app very easily if it grows. The idea behind this was - as much as possible - to have my cake and eat it too: get full Symfony, but get it small.

Anyways, I see your point - after all, this class is quite small - but that's my vote. We could move it into `FrameworkBundle` if we want to hide it a bit from the core components (especially since it requires `FrameworkBundle`.

**Comment by weaverryan** [comment]
Also, I don't envision that you'll normally have *just* one file, though that ability is cool. More that it's easy to create a minimal application easily: a kernel, controller classes, services and some configuration files if you choose.

File: src/Symfony/Component/HttpKernel/MicroKernel.php (hunk 1 above)
Comment (nicolas-grekas) [inline src/Symfony/Component/HttpKernel/MicroKernel.php]: I like ConfigurableKernel

**Comment by WouterJ** [comment]
I don't like the idea of using this in the Standard Edition. The Standard Edition should be the base template for common Symfony applications, which tend to be quite advanced. This still is a nice feature, but for 20% or so of the Symfony developers.

What about creating a new package, like `symfony/micro`, which requires `symfony/symfony` and this class. Then you can say "Require the `symfony/micro` package, use the `MicroKernel` class and you're ready".

**Comment by weaverryan** [comment]
What about inside `FrameworkBundle`? Having a new package for this one class looks like overkill, and this is just one small class.

**Comment by ogizanagi** [comment]
:+1: for `FrameworkBundle`

**Comment by weaverryan** [comment]
Is it too late to have this class considered to be merged?

**Comment by fabpot** [comment]
Honestly, I'm not very sure we need to include this class in Symfony. As you can see, the amount of code is minimal and advocating several ways of doing the same thing is always bad. Also, I don't see much benefits in using this instead of what we already have.

So, before merging this, I would want to understand the real use cases of this and why it's better than the current way. We can even decide to make it the preferred/only way at some point.

To sum up, as the code is minimal and because we can easily write a cookbook, I don't want to merge it now for 2.8/3.0.

**Comment by weaverryan** [comment]
@fabpot Ok, let me state my case! :).

Yes: there should only be one way to do things, and I've already removed a lot from the "micro kernel" idea for this. That's why I think this should be used in the SE. 

**1) Why would this be the preferred way?**

Here's the diff to make this the kernel for the SE: https://gist.github.com/weaverryan/25351e929a5a0d04c1e6/revisions#diff-237b910f44f219a85dbcda0e21295717. (it could be even smaller - if I didn't remove `routing_dev.yml`)

So why is this a better way?

A) Container loading is the same as now + you have access to the ContainerBuilder.

B) Route loading, same as now, but your code is in PHP, so you can use some logic, e.g. loading
extra routes in dev/test environments, or building routes in PHP.

This could be enough to accept this class into core.

**2) Use-cases**

>
A) Creating a micro-service or a demo with minimal files

In this kernel, the loading of container/routing resources is done very explicitly. So it would be very natural to start with the SE (or start from scratch) and create an app with less configuration files / directories. But, it would also still look *very* similar to a traditional SE project: everything runs through `configure***()` methods in both cases, which are explicit and clear in their purpose.

The setup for the service route loader and closure container loader are boilerplate, and making the
user go through this makes doing this less attractive, especially for beginners:, since the setup is quite advanced. But, if it's available as a core class, looking at the source code later could be used for teaching.

>
B) Clarity for learning.

I first teach that Symfony is a routing/controller/response system. What simpler way than to show them the `configureRoutes()` method with a simple PHP route first that points to a controller? Then, we could naturally refactor to loading configuration files. The same goes for the container: I explain that we "teach" the container how to instantiate our services. Doing that in a method called `configureServices()` using native PHP at first would be very natural. Then we can load a YAML file (from that same method) as a different way of organization.

Again, if you require the route/container loader setup, that'll be a problem. Even if I did this before the class for them, the code is still sitting there, exposing complexity I have to tell them to ignore. And, their project wouldn't look like the SE - eventually we'd have to refactor this back to the "normal" method.

Thanks!

**Comment by javiereguiluz** [comment]
Another use case: create a microkernel to use it in functional tests of Symfony bundles.

**Comment by jakzal** [comment]
> What about inside FrameworkBundle? Having a new package for this one class looks like overkill, and this is just one small class.

I wanted to write that this class could be used without the FrameworkBundle, but then I noticed it uses the framework extension. In which case the FrameworkBundle is probably the right place for it (unless we can make that extension is loaded optionally). Either way, I'd love to see this in core. I can see myself using it for simple projects, prototypes, workshop applications, or as Javier mentioned functional tests.

How about symfony/micro-edition?

**Comment by Pierstoval** [comment]
I totally agree with @weaverryan , this code should be in the core. It'd be an awesome step for a "Symfony nude/empty/micro/small/light Edition" that would require only the FrameworkBundle and a single Controller class.

File: src/Symfony/Component/HttpKernel/MicroKernel.php (hunk 1 above)
Comment (Pierstoval) [inline src/Symfony/Component/HttpKernel/MicroKernel.php]: ConfigurableKernel is cool :+1: 
 
What about StandardKernel or SimpleKernel ? Even though I'd vote for ConfigurableKernel :)

**Comment by henrikbjorn** [comment]
:+1: i like this idea as it does a lot of the stuff that i am trying to tie into Silex eg. the Routing.

File: src/Symfony/Component/HttpKernel/MicroKernel.php (hunk 1 above)
Comment (javiereguiluz) [inline src/Symfony/Component/HttpKernel/MicroKernel.php]: If this is "the" ConfigurableKernel ... does that mean that the other kernels are *not* configurable?

Comment (Pierstoval) [inline src/Symfony/Component/HttpKernel/MicroKernel.php]: This Kernel is abstract. Why not `AbstractKernel` then?

**Comment by Pierstoval** [comment]
I reviewed the code a bit, and actually, as this kernel is abstract, I think it could be good for the different `configure*` methods to be abstract too.
There are three methods to be overriden, for routes, services and extensions. 
With the `registerBundles` function, it means that there will be at least 4 methods defined in application kernel. If this abstract kernel is used, I bet it will be unlikely that the 3 abstracted methods would be empty. And if they're empty, it's not a problem: there are many small apps where the `services.yml` file is empty, or `routing.yml` contains only the `AppBundle/Controller/` annotation loader for every controller.
 
It'd allow a better start to understand how a kernel can be configured this way IMO.

**Comment by mfdj** [comment]
I like how flat this Kernel is (in the [PEP20](https://www.python.org/dev/peps/pep-0020/) sense)… `FlatKernel` is probably obtuse but it resonates with the design characteristic.

A great reason to have this in the core is it could make demonstrating the framework to a non-Symfony (or even non-PHP) developer. An app powered by a single file  — with no disclaimers about non-standard approach! — very appealing to the uninitiated.

File: src/Symfony/Component/HttpKernel/MicroKernel.php (hunk 1 above)
Comment (WouterJ) [inline src/Symfony/Component/HttpKernel/MicroKernel.php]: Kernel is abstract as well...

**Comment by fabpot** [comment]
I've just made some changes:

 * moved the class to `FrameworkBundle`;
 * merged the `configureServices()` and `configureExtensions()` into one new `configureContainer()` method;
 * made the two `configure*()` method abstract.

"We" want to merge that into 2.8, so ping @symfony/deciders

**Comment by fabpot** [comment]
:+1:

**Comment by fabpot** [comment]
And now with some tests \o/

**Comment by dunglas** [comment]
:+1:

**Comment by Pierstoval** [comment]
Yeah, trick or treat!
 
Can't wait for this kernel to be alive in 2.8, it'd be a great move, and I bet there will be many "Symfony micro-apps" in the future :)

File: src/Symfony/Bundle/FrameworkBundle/Kernel/MicroKernel.php [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+    {
+        $loader->load(function (ContainerBuilder $container) use ($loader) {
+            $container->loadFromExtension('framework', array(
+                'router' => array(
+                    'resource' => 'kernel:loadRoutes',
+                    'type' => 'service',
+                ),
+            ));
+
+            $this->configureContainer($container, $loader);
```
Comment (xabbuh) [inline src/Symfony/Bundle/FrameworkBundle/Kernel/MicroKernel.php]: `$this` is not available here.

File: src/Symfony/Bundle/FrameworkBundle/Kernel/MicroKernel.php [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+
+            $container->addObjectResource($this);
```
Comment (aitboudad) [inline src/Symfony/Bundle/FrameworkBundle/Kernel/MicroKernel.php]: same here

**Comment by aitboudad** [comment]
:+1:

File: src/Symfony/Bundle/FrameworkBundle/Kernel/MicroKernel.php (hunk 5 above)
Comment (fabpot) [inline src/Symfony/Bundle/FrameworkBundle/Kernel/MicroKernel.php]: fixed

File: src/Symfony/Bundle/FrameworkBundle/Kernel/MicroKernel.php (hunk 6 above)
Comment (fabpot) [inline src/Symfony/Bundle/FrameworkBundle/Kernel/MicroKernel.php]: fixed

**Comment by fabpot** [comment]
Making this class compatible with PHP 5.3 is a PITA. What about making it require PHP 5.5?

**Comment by weaverryan** [comment]
@fabpot is there something beyond the ugly $that? I don't think requiring 5.5 will cause many practical issues, but the $that doesn't seem like a deal breaker (and we could change it in 3.0 immediately). But if there is some other issue, then 5.5 is ok - I just want it to be in :)

**Comment by fabpot** [comment]
`$that is not enough`, the method should also be changed to public instead of protected.

**Comment by nicolas-grekas** [comment]
I personally would be ok if we say that this new feature in 5.4+ only.
What about replacing this abstract class by a trait + an interface btw?
It would naturally enforce the 5.4+ requirement, and also provide a cleaner type-system.

**Comment by aitboudad** [comment]
We can solve it by adding another public method and mark it as internal
```php
/**
 * This method is added because it's needs to be callable from a closure in PHP 5.3. It should be removed in 3.0.
 * @internal
 */
public function configure(ContainerBuilder $c, LoaderInterface $loader)
{
    $this->configureContainer($c, $loader);
}
```

**Comment by GromNaN** [comment]
I agree with @nicolas-grekas on making a trait instead of a class for this feature. An interface is not necessary as traits can define abstract methods.

```php
class AppKernel extends Kernel
{
    use MicroKernelTrait;

    protected function configureRoutes(RouteCollectionBuilder $routes) {}
    protected function configureContainer(ContainerBuilder $c, LoaderInterface $loader) {}
}
```

**Comment by weaverryan** [comment]
The trait would be clean, but I don't think traits are always clear. In this case, I may still think I need to implement registerContainerConfiguration in AppKernel (and if you do that, you'll kill the functionality). But maybe no one else would think in this way.

Since this is an "extra" feature that someone on 5.3 could just duplicate and reimplement anyways, I don't see an issue with not supporting 5.3... other than getting the tests to pass. Is it possible to skip a test for a Php version? Or are we making more trouble?

**Comment by nicolas-grekas** [comment]
If copy/pasting is fine, then it means semantically you need a trait. Having an abstract class creates a new type that anyone could hint in some signature, which would be a mistake. That's what trait are for. 

> I may still think I need to implement registerContainerConfiguration in AppKernel

then read the doc again about traits... (not you @weaverryan , but that hypothetical "I")

**Comment by weaverryan** [comment]
A trait is really fine with me :)

**Comment by Pierstoval** [comment]
Actually, a trait sounds more like a "behavior" to me more than a proper application of a concept. An abstract class, instead, looks more than a working example of a class that only needs some configuration (= the abstract methods) to work properly when using it.
 
Indeed, using a trait would force the PHP requirement to be 5.4+ but is this really needed?

File: src/Symfony/Bundle/FrameworkBundle/Kernel/MicroKernel.php [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+     * @param LoaderInterface  $loader
+     */
+    abstract protected function configureContainer(ContainerBuilder $c, LoaderInterface $loader);
+
+    /**
+     * {@inheritdoc}
+     */
+    public function registerContainerConfiguration(LoaderInterface $loader)
+    {
+        $that = $this;
```
Comment (Pierstoval) [inline src/Symfony/Bundle/FrameworkBundle/Kernel/MicroKernel.php]: Why not simply rename `$that` into `$kernel`? It seems like a dirty workaround to me, whereas `$kernel` is more concise, even if "it's the same".

Comment (weaverryan) [inline src/Symfony/Bundle/FrameworkBundle/Kernel/MicroKernel.php]: it is a dirty workaround :). The worse problem is that `configureContainer` would need to be public for 5.3

Comment (Pierstoval) [inline src/Symfony/Bundle/FrameworkBundle/Kernel/MicroKernel.php]: Oh sorry, bly me I did not realized that the problem came from the `configureContainer` visibility :confused:

**Comment by nicolas-grekas** [comment]
:+1:

**Comment by weaverryan** [comment]
:+1:

**Comment by dunglas** [comment]
:+1: the trait looks good.

**Comment by weaverryan** [comment]
Thanks Fabien :)

**Comment by Pierstoval** [comment]
:+1:

**Comment by henrikbjorn** [comment]
While playing around with it, i found it was hard to get to import another routing file using RouteCollectionBuilder, because its `import()` function return a new builder.

Is there a better way of accomplishing this (notice the `&` which replaces the builder):

``` php
    protected function configureRoutes(RouteCollectionBuilder &$routes)
    {
        $routes = $routes->import($this->rootDir . '/config/routing.yml', null);
    }
```

**Comment by GromNaN** [comment]
The method `configureRoutes` could have to return the `$routes`.

**Comment by weaverryan** [comment]
@henrikbjorn thanks for playing with it. Here is how you're *supposed* to import routes, but there could be a usability problem (as you are quite smart and this wasn't obvious to you):

```
$routes->mount('/', $routes->import(__DIR__.'/config/routing.yml');
```

In other words, it *does* create a new `RouteCollectionBuilder`, but you still need to mount it into the existing builder.

**Comment by henrikbjorn** [comment]
@weaverryan assume i have several routing files i want added, if both  have the same prefix i assume the latter will override the former?

**Comment by weaverryan** [comment]
@henrikbjorn it uses the same mechanism as normal route importing. So if you import 5 routes from `admin.yml` and 5 routes from `admin2.yml` - both with the prefix `/admin`, and you will end up with 10 routes. The only time later routes would override earlier routes is if they have the same route name - the prefix won't matter with that.

**Comment by WouterJ** [comment]
Maybe we should allow mounting without a prefix?

**Comment by weaverryan** [comment]
Before I answer that, what is the *problem* with the current implementation? :) I have some small things that I don't think are necessarily perfect, but I only want to change things if people like @henrikbjorn say "Hey, I was confused by X, it would have been more clear if I could just call Y".

**Comment by WouterJ** [comment]
Let's describe what I think (as a completely non-native) the discussed function names mean:

 * `mount()` bind 2 things (first and second argument) together
 * `import()` includes something into the current stack

It turns out `mount()` doesn't mount argument 1 and 2, it mounts argument 2 in the current route collection with an optional prefix defined by 1.

Furthermore, it turns out `import` actually doesn't do anything more than loading and returning what it just loaded. The thing you loaded now has to be attached/mounted to the main route collection.

**Comment by weaverryan** [comment]
About `mount()`, it's exactly as it is in Silex (http://silex.sensiolabs.org/doc/organizing_controllers.html). So if it hasn't been a problem there for people, we can assume it won't be a problem here (if it has been a problem, we should't repeat it).

About `import()`, I share your concerns. However, if you automatically put the new routes into the `RouteCollectionBuilder`, then you're not able to modify then further - e.g. add defaults or requirements to  the routes in *just* that collection. Again, this idea comes from Silex (http://silex.sensiolabs.org/doc/providers.html#controller-providers), though it's a bit more obvious there, because you create new class, so it's obvious that it's not being imported until you actually add it to your app. Is it as simple as calling `import()` something else? Or making `import()` actually *import* and adding a less-used method that returns the new `RouteCollectionBuilder` without mounting it?

**Comment by WouterJ** [comment]
We can rename it to `load()` or import the route collection and return the imported route collection, allowing to change the instance. (this is done with registering Container definitions for instance)

**Comment by weaverryan** [comment]
@WouterJ you're right - I didn't think about that - that's great :). See #16477.

**Comment by henrikbjorn** [comment]
@weaverryan i think the confusing part for me was the `$prefix` in mount is the first argument, and i would the assume it would be mandatory and therefor override my ealier prefix. But if the argument order was different eg `mount($routes, $prefix = null)` then i would make sense.

But that would proberly be a bigger BC break than your proposed solution.

**Comment by janit** [comment]
All I see is references to 2.8. Will this be in 3.0 too?

**Comment by Pierstoval** [comment]
It should be merged in the next "merge 2.8 on 3.0" I guess :)

**Comment by SoboLAN** [comment]
@fabpot I see this was merged in **2.8** branch. But 2.8 requires **PHP 5.3.9**, while traits are only available in **PHP 5.4+** .

Are you sure it's correct to have this in **2.8** ?

**Comment by WouterJ** [comment]
@SoboLAN this is an additional feature. Symfony doesn't require you to use this feature, so the minimum requirement is still 5.3.9.

If you want to use this feature, you have to use PHP 5.4.

**Comment by SoboLAN** [comment]
@WouterJ You could say that about most of the functionality provided by Symfony.

If a minimum PHP version is specified, then I expect that absolutely everything will run without problems if I use Symfony on that PHP version (or newer). That's the whole point of why Composer allows specifying something like this.

I think it's a very reasonable assumption. Not respecting that is kind of alarming for me because I can never know what to expect. I would constantly ask myself:

> If I use this method in my project, will it work in production servers ? (where I have a locked and usually older PHP version than on my dev machine)

If you (the deciders of Symfony) strongly believe that it should be like this, then fine. Either way I have no authority to dictate anything around here. But think about it: if you do this now, then where do you draw the line ? If you go down this path, where do you stop ?

I think that it's a dangerous precedent because it takes away all the confidence in that `requires PHP version X` statement.

Without that confidence, what is a developer or devops going to do ? What if he has a big Symfony project (or 20 projects) in his production systems ? Test absolutely every functionality when he makes a patch update (e.g. 2.6.9 -> 2.6.10) ? Just to be sure that what is written in `composer.json` is not a lie and that his website won't blow up ?

**Comment by WouterJ** [comment]
There are many packages in the `suggest` sections of Symfony's framework, components and bundles. These are all libraries that have some kind of integration in the component/bundle, but aren't required for the main usage (for instance, `doctrine/annotations` is required when using annotations with the Validator component, but as you can use Yaml/Xml/PHP, it's in the `suggest` section).

Also, 95% of the people that are going to use Symfony are not going to use this trait. It's a really nice feature, but more to advocate Symfony (*"Symfony isn't that massive huge framework you think it is!"*) and for some people that are not going to base on top of the Standard Edition (and most people installing 2.8 will stick with the Standard Edition).

At last, it's very clear (or debuggable within 30 seconds) that this feature requires PHP 5.4 as you need to use the trait yourself. A developer using this feature on a PHP 5.3 server will either know the trait feature and identifies it as something that requires PHP 5.4 or will get an error message, search for "php trait" on google and find out that it requries PHP 5.4. Furthermore, I'm sure the documentation about this feature will mention the PHP 5.4 requirement "above the fold" of the article.

So for these reasons, I think it's safe to put this PHP-5.4-requiring-feature in Symfony 2.8. I don't think you need to be afraid that Symfony will cross the line many times in the future. The Symfony framework has a core team that's very focussed on it's usage (with the BC promise, easy upgrade paths, etc. as results).

At last, I want to thank you for sharing your worries in such a clear and long comment!

**Comment by weaverryan** [comment]
I agree with Wouter - this is a very special situation - so don't worry @SoboLAN, we take this thing quite seriously.

Btw, nothing depends on this class in Symfony, it's a total extra feature. If you want it on 5.3, you can safely copy this class and tweak it to work with 5.3.

Cheers!

**Comment by Pierstoval** [comment]
Plus, to add another argument, this feature is mostly for new projects, and I doubt that developers creating new projects will use an old PHP version.

**Comment by psampaz** [comment]
I totally agree with @SoboLAN.

**Comment by javiereguiluz** [comment]
@WouterJ @weaverryan what you say is true and reasonable. But at the same time I find @SoboLAN's comments very reasonable too. If some code claims 5.3 compatibility, I understad that it's for its entire codebase.

**Comment by Pierstoval** [comment]
[gedmo/doctrine-extensions](https://github.com/Atlantic18/DoctrineExtensions) has a PHP requirement of `>=5.3.2` and it has traits. If you don't want to use them, you don't have to, and your code won't break.
 
It's been already said:

* 95% of the people that are going to use Symfony are not going to use this trait
* nothing depends on this class in Symfony, it's a total extra feature
* this feature is mostly for new projects

For me, these reasons are enough to keep it in the 2.8 branch

**Comment by Tobion** [comment]
Btw we already had traits in symfony before (ContainerAwareTrait). Not sure what the goal of this discussion is. Nobody forces you to use it.

**Comment by SoboLAN** [comment]
@Pierstoval @Tobion I totally understand what you're saying. No one is making you use these features.

But the problem in these situations is that you may use some classes/functionalities without knowing it.

Let's take a simple example and say that I have the following line in one of my Controllers:

```php
$this->get('validator')->validate($someObject, array('group1', 'group2'));
```

I'm not using what I'm not supposed to. But, behind the scenes, how much is going on in order to validate that object ? You have to:

- build the validator object (which is very simple)
- let's say read a `validation.yml` file
- interpret the file
- have the Validator go through it and determine which constraints contain those specified groups
- apply those constraints
- build violation errors
- etc.

There are probably thousands of lines of code being executed inside that 1 line from above.

If somewhere in all that code you add code (like for example a trait) that is not compatible with what PHP version you specify in composer, then my project will maybe fail, usually in ways that can be difficult to reproduce.

@Pierstoval @Tobion I only presented above the reason for why this whole discussion started.

@WouterJ assured me that something like this would never happen. I hope he's right. I'm very familiar with Symfony's promise to have excellent backwards-compatibility. I will just have faith that bad scenarios like the one above will never happen :) .

I hope I did not piss people off with my concerns. If I did, it was not my intention. I was just somewhat worried.

**Comment by nicolas-grekas** [comment]
@SoboLAN Symfony does not rely on the MicroKernelTrait, which means it's true that Symfony works with 5.3. No project is going to break because we added the trait.
If someone or some lib start relying on the trait to provide some feature, then it's their responsibility to enforce the correct minimum PHP version.
